SpiceSharp/Simulations/IntegrationMethods/Spice/Gear/GearStateDerivative.cs

```diff
@@ -0,0 +1,148 @@
+
+            /// <summary>
+            /// Gets or sets the current value
+            /// </summary>
+            public override double Current
+            {
+                get => _states[0].State[_index];
+                set => _states[0].State[_index] = value;
+            }
+
+            /// <summary>
+            /// Gets a previous state value
+            /// </summary>
+            /// <param name="index">Number of points to go back</param>
+            /// <returns></returns>
+            public override double this[int index] => _states[index].State[_index];
+
+            /// <summary>
+            /// Gets the derivative of the state
+            /// </summary>
+            public override double Derivative => _states[0].State[_index + 1];
+
+            /// <summary>
+            /// Constructor
+            /// </summary>
+            /// <param name="method">Parent integration method</param>
+            public GearStateDerivative(Gear method)
+            {
+                _method = method;
+                _index = method.StateManager.AllocateState(1);
+                _states = method.IntegrationStates;
+            }
+
+            /// <summary>
+            /// Calculate the jacobian
+            /// </summary>
+            /// <param name="derivative">Derivative</param>
+            /// <returns></returns>
+            public override double Jacobian(double derivative) => derivative * _method.Slope;
+
+            /// <summary>
+            /// Calculate the RHS-vector contribution for linear components
+            /// </summary>
+            /// <returns></returns>
+            public override double RhsCurrent() =>
+                _states[0].State[_index + 1] - _method.Slope * _states[0].State[_index];
+
+            /// <summary>
+            /// Calculate the derivative
+            /// </summary>
+            public override void Integrate()
+            {
+                var derivativeIndex = _index + 1;
+                var ag = _method.Coefficients;
+
+                var current = _states[0].State;
+                current[derivativeIndex] = 0.0;
+                for (var i = 0; i <= _method.Order; i++)
+                    current[derivativeIndex] += ag[i] * _states[i].State[_index];
+            }
+
+            /// <summary>
+            /// Truncate the time step
+            /// </summary>
+            /// <returns></returns>
+            public double Truncate()
+            {
+                var derivativeIndex = _index + 1;
+                var current = _states[0].State;
+                var previous = _states[1].State;
+
+                var diff = new double[_method.MaxOrder + 2];
+                var deltmp = new double[_states.Length];
+
+                // Calculate the tolerance
+                var volttol =
+                    1e-12 + 1e-3 * Math.Max(Math.Abs(current[derivativeIndex]), Math.Abs(previous[derivativeIndex]));
```

Comment:
@svenboulanger Please provide consts for 1e-12 or 1e-3 or a method to calculate tol.

Reply:
👍